### ISSUE TITLE
A way to check the appmenu-gtk-module version

### ISSUE THREAD

**Issue by lestcape** [rilian-la-te/vala-panel-appmenu#234]
As appmenu-gtk-module can change between releases in the same way of all other things that are integrate here, is not a problem to vala-panel-appmenu at all. But if an external application want to use the appmenu-gtk-module, need to be a way to detected the installed version to act in concordance with any change that will be made inside the appmenu-gtk-module.

**Comment by rilian-la-te** [comment]
If it is built together, version exists in name, like libapppmenu-gtk-parser.so.0.6.94

**Comment by rilian-la-te** [comment]
If it is not, I can integrate a same mechanic. Also there is appmenu-gtk-parser.pc pkg-config file.

**Comment by rilian-la-te** [comment]
3. Just cat (or ls) a file and check for version.
Why you do not want this? Can you describe disadvantages of this variant?

**Comment by lestcape** [comment]
Yes, the shell have only one thread after this specific version, that now have multitasking. spaw a command line is pretty wrong inside the shell.

**Comment by rilian-la-te** [comment]
We can use a Gio.File, not command line.

**Comment by lestcape** [comment]
You will embed a query inside Gio.File? How?

**Comment by rilian-la-te** [comment]
Get name of Gio.File, and then strip a part before .so and .so itself.

**Comment by rilian-la-te** [comment]
AFAIK, GFile is following symlinks.

**Comment by lestcape** [comment]
Seriously, where is the file path? I need to localize the file...

**Comment by rilian-la-te** [comment]
/usr/lib/<triplet if debian>.

**Comment by rilian-la-te** [comment]
Same place where you search for gtk+ modules, but strip gtk+-3.0/modules

**Comment by lestcape** [comment]
Ok. Actually there are not version in this libraries.

![captura de pantalla de 2018-05-06 14-55-35](https://user-images.githubusercontent.com/1389572/39677266-99106a50-513d-11e8-85cc-9ec3591df243.png)

**Comment by lestcape** [comment]
Where is the file appmenu-gtk-parser.pc??
Actually, i don't need a version. Have a version is just for ensure i will be able to support several versions at the same time. Not all distros are rolling release, so i can have the same gnome shell version using different appmenu-gtk-module version and my extension need to work with both, not matter the change you will do.

**Comment by rilian-la-te** [comment]
/usr/lib/"maybe debian triplet"/pkgconfig/appmenu-gtk3-parser.pc

**Comment by rilian-la-te** [comment]
appmenu-gtk-module functionality will not be dependent to version, because I need to support all standard protocols on all DEs. And it should be backward-compatible with unity-gtk-module. 
If I will add a custom protocol, I will integrate version info into it.
For now, it should work transparently for users - they install it and no action required.

**Comment by lestcape** [comment]
I can not localize this file and also if can be installed in different locations, this is not an option.

If you add or remove gsettings key to active or deactivated behaviours between versions, how i will be sure what i need to do?

https://github.com/rilian-la-te/vala-panel-appmenu/commit/28677d45ba23d1fc28cebb836ea85235015c2db6

If a user have gnome shell and also mate and if the user modify this key (default value false) in mate, what will be happen in the gnome-shell version? Remember that this is just an example.

**Comment by rilian-la-te** [comment]
This file can be installed via devel package (on Arch no devel packages exists).
About this strange user - I do not know how to deal with him. 


For #196  I need to provide a way to not remove a menu when there is only HUD. I can do it by following ways:
1. Add a custom protocol, which interacts with module and send registrar role to it. But it breaks backward compatibility with unity-gtk-module and another realizations.
2. Add an environment variable - I do not like it and there can be issues when variable is reset by user or by terminal.
3. Add a gsettings key - most easier.
4. Maybe another way than I do not see for now.

But for 3 we need to deal with multi-DE issues.

**Comment by lestcape** [comment]
Yes. this is the point, sorry for the wrong link : https://github.com/rilian-la-te/vala-panel-appmenu/issues/196#issuecomment-381442195

I don't want to over complicate the appmenu-gtk-module and i think is fine if will be used for more things... So, i just want a way to act in concordance. For that reason, i think the most easy way is have a version to be checked.

**Comment by rilian-la-te** [comment]
You need a module version or schema version or more specific version?
For example, Wayland fix can also change appmenu-gtk-module Wayland protocol.

**Comment by rilian-la-te** [comment]
I will think about it (about storing some environment info).

**Comment by lestcape** [comment]
Yes, is this or delegate properties to the enviroment and check the existence of the gsettings of this specific enviroment with is really create a lot of possible options. I don't think that this will be a good idea anyway.

**Comment by lestcape** [comment]
Do you see if this help for Firefox support outside ubuntu? https://launchpad.net/ubuntu/xenial/+package/firefox-globalmenu

**Comment by rilian-la-te** [comment]
Nope. It is just a stub package now.

**Comment by lestcape** [comment]
But in theory this package contain the things that need to be changed... I remember someone write a firefox plugins to export the global menu, but probably won' t work right now.

**Comment by rilian-la-te** [comment]
It won't work, because WebExtensions AFAIK does not give access to menu to extensions.

**Comment by lestcape** [comment]
This is the ubuntu path: https://aur.archlinux.org/cgit/aur.git/tree/unity-menubar.patch?h=firefox-kde-opensuse

**Comment by rilian-la-te** [comment]
It is) But it will not be accepted to mainline, so, dead end.

**Comment by lestcape** [comment]
This is one old firefox extension: https://gitorious.org/firefox-gnome-globalmenu/firefox-gnome-globalmenu?p=firefox-gnome-globalmenu:firefox-gnome-globalmenu.git;a=shortlog;h=refs/heads/master

**Comment by rilian-la-te** [comment]
I know about it too. For me, firefox is impossible without upstream support. I cannot hijack to firefox, because it does not use GTK in its menus.
